rails_event_store/spec/instrumented_repository_spec.rb

```diff
@@ -0,0 +1,176 @@
+      end
+    end
+
+    describe "#has_event?" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+
+        instrumented_repository.has_event?(42)
+
+        expect(some_repository).to have_received(:has_event?).with(42)
+      end
+    end
+
+    describe "#last_stream_event" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+
+        instrumented_repository.last_stream_event("SomeStream")
+
+        expect(some_repository).to have_received(:last_stream_event).with("SomeStream")
+      end
+    end
+
+    describe "#read_event" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+
+        instrumented_repository.read_event(42)
+
+        expect(some_repository).to have_received(:read_event).with(42)
+      end
+
+      specify "instruments" do
+        some_repository = double
+        allow(some_repository).to receive(:read_event)
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+        notification_calls = subscribe_to("read_event.repository.rails_event_store")
+
+        instrumented_repository.read_event(42)
+
+        expect(notification_calls).to eq([
+          { event_id: 42 }
+        ])
+      end
+    end
+
+    describe "#read" do
+      specify "wraps around original implementation" do
+        some_repository = spy
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+        specification = double
+
+        instrumented_repository.read(specification)
+
+        expect(some_repository).to have_received(:read).with(specification)
+      end
+
+      specify "instruments" do
+        some_repository = double
+        allow(some_repository).to receive(:read)
+        instrumented_repository = InstrumentedRepository.new(some_repository)
+        notification_calls = subscribe_to("read.repository.rails_event_store")
+        specification = double
+
+        instrumented_repository.read(specification)
+
+        expect(notification_calls).to eq([
+          { specification: specification }
+        ])
+      end
+    end
+
+    def subscribe_to(name)
+      received_payloads = []
+      ActiveSupport::Notifications.subscribe(name) do |_name, _start, _finish, _id, payload|
+        received_payloads << payload
+      end
+      received_payloads
+    end
+  end
+end
+
+module RailsEventStore
+  RSpec.describe InstrumentedRepository do
+    subject do
+      InstrumentedRepository.new(RubyEventStore::InMemoryRepository.new)
+    end
+
+    let(:test_race_conditions_any)   { false }
+    let(:test_race_conditions_auto)  { false }
+    let(:test_expected_version_auto) { false }
+    let(:test_link_events_to_stream) { false }
```

Comment:
why `false` here as well ?

Reply:
Dunno about these, I've initially skipped these values because they've required some additional methods to be defined. I will look at it.

Reply:
It doesn't make sense probably to test race conditions but it does make sense to verify that linking events still works correctly, don't you think?